[export] make aot_export_module uses dynamo's fake_mode

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113681

Fixes #110100 by making aot_export_modules uses dynamo.export's fake_mode in export.

Test Plan:
Add new tests. One of the test places the fake tensor on cuda devices manually and we are able to export the program and preserve the device information in the final produced graph module even on a machine that installs a cpu version of pytorch. One workaround we need to do is to set all tensor's requires_grad to false as fake tensor with cuda devices doesn't compose well with aot_autograd right now.


cc @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo

Differential Revision: [D51414973](https://our.internmc.facebook.com/intern/diff/D51414973)

ydwu4 committed:
[export] make aot_export_module uses dynamo's fake_mode

[ghstack-poisoned]

zhxchen17 commented on torch/_export/__init__.py:
```diff
@@ -563,7 +563,17 @@ def export(
         preserve_module_call_signature=preserve_module_call_signature,
     )
 
+def disable_prexisiting_fake_mode(fn):
```
```suggestion
def _disable_prexisiting_fake_mode(fn):
```

zhxchen17 reviewed (commented):
Nice! Could you also link the issue number in the PR summary so we can get a reference there?

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -563,7 +563,17 @@ def export(
         preserve_module_call_signature=preserve_module_call_signature,
     )
 
+def disable_prexisiting_fake_mode(fn):
```
Does this mean we are always using dynamo produced fake mode?

ydwu4 replied:
yes! dynamo's fake_mode will be the single source of truth. This is easier to reason about and it also makes sure we won't mutate user provided fake_mode. The user provided input and model.named_parameters and buffers are still using outer fake_mode.

tugsbayasgalan commented on torch/_export/__init__.py:
```diff
@@ -695,11 +706,12 @@ def _export(
-        (*fake_args, *_reorder_kwargs_by_names(orig_args, fake_args, fake_kwargs).values()),
-        trace_joint=False
-    )
+    with torch.nn.utils.stateless._reparametrize_module(gm_torch_level, fake_params_buffers):
```
Just curious, why is this needed?

ydwu4 replied:
The gm_torch_level.named_parameters().fake_mode is the outer fake_mode because dynamo doesn't wrap them as fake tensors (unlike what it does for inputs) and register the original nn_modules as it is to gm_torch_level. 

This reparametrize_module make sure inputs and gm_torch_level.params/buffers have the same fake_mode, otherwise aot_export_module will error out because it sees a mix of fake_modes.

SherlockNoMad replied:
>This reparametrize_module makes sure inputs and gm_torch_level.params/buffers have the same fake_mode, otherwise aot_export_module will error out because it sees a mix of fake_modes. And we want aot_export_module to use the fake_tensor mode in dynamo to keep the pipeline easy to reason about.

nit, add this as comment.

ydwu4 committed:
Update on "[export] make aot_export_module uses dynamo's fake_mode"


Fixes #110100 by making aot_export_modules uses dynamo.export's fake_mode in export.


cc avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo

[ghstack-poisoned]

ydwu4 committed:
Update on "[export] make aot_export_module uses dynamo's fake_mode"


Fixes #110100 by making aot_export_modules uses dynamo.export's fake_mode in export.

Test Plan:
Add new tests. One of the test places the fake tensor on cuda devices manually and we are able to export the program and preserve the device information in the final produced graph module. One workaround we need to do is to set all tensor's requires_grad to false as fake tensor with cuda devices doesn't compose well with aot_autograd right now.


cc avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo

[ghstack-poisoned]

ydwu4 committed:
Update on "[export] make aot_export_module uses dynamo's fake_mode"


Fixes #110100 by making aot_export_modules uses dynamo.export's fake_mode in export.

Test Plan:
Add new tests. One of the test places the fake tensor on cuda devices manually and we are able to export the program and preserve the device information in the final produced graph module. One workaround we need to do is to set all tensor's requires_grad to false as fake tensor with cuda devices doesn't compose well with aot_autograd right now.


cc avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo

[ghstack-poisoned]

SherlockNoMad commented on test/export/test_export.py:
```diff
@@ -1548,6 +1548,75 @@ def forward(self, x):
+            x = torch.rand(5, 2, 2)
+            model = Model()
+
+        # Mannuly se the fake_device of fake tensors.
```
set

SherlockNoMad commented on test/export/test_export.py:
```diff
@@ -1548,6 +1548,75 @@ def forward(self, x):
+                out = self.linear(x)
+                return out
+
+        # Put the inputs on a device
+        with fake_mode, torch.device('meta'):
+            x = torch.rand(5, 2, 2)
+            model = Model()
+
+        # Mannuly se the fake_device of fake tensors.
+        x.fake_device = torch.device('cuda:0')
+        for n, p in model.named_parameters():
+            p.fake_device = torch.device('cuda:0')
+
+        # Need to set all the requires_grad of tensors to False, because fake_tensor with CUDA device
+        # doesn't quite work well with aot_autograd right now due to some logic fails
+        # the check in call getDeviceGuardImpl in InputMetadata.
+        x.requires_grad = False
+        for n, p in model.named_parameters():
+            p.requires_grad = False
+
```
I wonder if we can have a better UX for this workflow. 

I imagine something like this 

```
with fake_mode, torch.no_grad():
    x = torch.rand(2, device='cuda:0') # this sets x.fake_device under the hood
    model = Model()
    model.submodule1.to("cuda:0") # parameters are fakified and placed on "cuda:0"
    model.submodule2.to("cuda:1")
```

SherlockNoMad replied:
This mimics the normal model construction. 

In such case, use just need to add `with fake_mode`, and no other changes are needed.

ydwu4 replied:
Let me check whether the UX can work. with torch.no_grad() doesn't set all the param/buffer's requires_grad to False somehow. Haven't checked the to("cuda:0") thing.

SherlockNoMad reviewed (approved):
lgtm, and I wonder if we can make the UX better.

SherlockNoMad commented:
Second though, to make this really useable, user shouldn't need to manipulate x.fake_device. User should write x.to("real_device"), and under fakemode, we handle the fake device placement for them.

ydwu4 committed:
Update on "[export] make aot_export_module uses dynamo's fake_mode"


Fixes #110100 by making aot_export_modules uses dynamo.export's fake_mode in export.

Test Plan:
Add new tests. One of the test places the fake tensor on cuda devices manually and we are able to export the program and preserve the device information in the final produced graph module even on a machine that installs a cpu version of pytorch. One workaround we need to do is to set all tensor's requires_grad to false as fake tensor with cuda devices doesn't compose well with aot_autograd right now.


cc avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change, but it is failing an internal ExecuTorch test' -c ghfirst

The error is as follows:

```
======================================================================
FAIL: test_asr_encoders (pye.model_inventory.asr_models.milan_dictation.test_xnnpack_asr.TestXnnpackPartitioner)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/pye/model_inventory/asr_models/milan_dictation/test_xnnpack_asr.py", line 28, in test_asr_encoders
    exported_program = gen.get_exported_program(ModelVariant.FP32, methods[0])
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/pye/lib/EagerModelBase.py", line 164, in wrapper
    return deepcopy(fn(*args, **kwargs))
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/pye/model_inventory/asr_models/milan_dictation/MilanDictationModel.py", line 277, in get_exported_program
    exported_program = exir.capture(
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/executorch/exir/capture/_capture.py", line 171, in capture
    ep = export(f, args, constraints=constraints)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/torch/_export/__init__.py", line 549, in export
    return _export(
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/torch/_export/__init__.py", line 562, in wrapper
    return fn(*args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/torch/_export/__init__.py", line 615, in _export
    fake_args, fake_kwargs, fake_params_buffers, dynamo_fake_mode = _convert_input_to_fake(gm_torch_level, args, kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/5408a275d581d7c2/pye/model_inventory/asr_models/milan_dictation/__test_xnnpack_asr__/test_xnnpack_asr#link-tree/torch/_export/__init__.py", line 412, in _convert_input_to_fake
    assert fake_mode is not None, "Cannot find fake_mode attatched to the graph's placeholders."
AssertionError: Cannot find fake_mode attatched to the graph's placeholders.

----------------------------------------------------------------------
Ran 1 test in 129.417s

FAILED (failures=1)
```

ydwu4 committed:
Update on "[export] make aot_export_module uses dynamo's fake_mode"


Fixes #110100 by making aot_export_modules uses dynamo.export's fake_mode in export.

Test Plan:
Add new tests. One of the test places the fake tensor on cuda devices manually and we are able to export the program and preserve the device information in the final produced graph module even on a machine that installs a cpu version of pytorch. One workaround we need to do is to set all tensor's requires_grad to false as fake tensor with cuda devices doesn't compose well with aot_autograd right now.


cc avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo

[ghstack-poisoned]

ydwu4 commented:
Internal tests have a model that takes no inputs, this makes _convert_to_fake_input unhappy. We detect this case and set the fake_mode to None in this case. Also added a new test for it.

ydwu4 commented:
@ydwu4 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51414973).

clee2000 commented:
@pytorchbot merge -f "merged internally 6hrs ago, not sure why bot didn't work"

ydwu4 committed:
Update on "[export] make aot_export_module uses dynamo's fake_mode"


Fixes #110100 by making aot_export_modules uses dynamo.export's fake_mode in export.

Test Plan:
Add new tests. One of the test places the fake tensor on cuda devices manually and we are able to export the program and preserve the device information in the final produced graph module even on a machine that installs a cpu version of pytorch. One workaround we need to do is to set all tensor's requires_grad to false as fake tensor with cuda devices doesn't compose well with aot_autograd right now.


cc avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan angelayi suo

Differential Revision: [D51414973](https://our.internmc.facebook.com/intern/diff/D51414973)

[ghstack-poisoned]

thiagocrepaldi commented:
should this issue be closed now that #114009 was relanded?
